File: fastapi_server/src/models/audiobook.py
import asyncio
import arrow
from piccolo.columns import ForeignKey, Integer, Text, Timestamptz, Boolean, JSON
from piccolo.table import Table



# await AudiobookBook.create_table(if_not_exists=True)
# await AudiobookChapter.create_table(if_not_exists=True)
class AudiobookBook(Table, tablename="litestar_audiobook_book"):
    uploaded_by = Text(required=True)
    book_title = Text(required=True)
    book_author = Text(required=True)
    chapter_count = Integer(required=True)
    upload_date = Timestamptz(default=arrow.utcnow().datetime)
    custom_book_title = Text()
    custom_book_author = Text()
    # Soft delete flag
    deleted = Boolean(default=False)


class AudiobookChapter(Table, tablename="litestar_audiobook_chapter"):
    book = ForeignKey(references=AudiobookBook)
    queued = Timestamptz()
    started_converting = Timestamptz()
    chapter_title = Text(required=True)
    chapter_number = Integer(required=True)
    word_count = Integer(required=True)
    sentence_count = Integer(required=True)
    content = Text(required=True)
    minio_object_name = Text()
    audio_settings = JSON()

File: svelte_frontends/apps/audiobook/src/lib/types/audiobook.ts
// Types for Audiobook feature
import type { VoiceOption as ApiVoiceOption, QueueChapterRequest } from "@repo/api-types"

export type AudioSettings = QueueChapterRequest
export type VoiceOption = ApiVoiceOption

export interface AudiobookBook {
    id: number
    uploaded_by: string
    book_title: string
    book_author: string
    custom_book_title: string
    custom_book_author: string
    chapter_count: number
    upload_date: string
}

export interface AudiobookChapter {
    id: number
    book: number
    chapter_title: string
    chapter_number: number
    word_count: number
    sentence_count: number
    queued: string | null
    started_converting: string | null
    minio_object_name: string | null
    audio_settings: string | null
}

export interface AudiobookChapterQueryResult {
    id: number
    book_id: number
    number_in_queue: number | null
    is_converting: boolean
    has_audio: boolean
    chapter_title: string
    chapter_number: number
    sentence_count: number
    minio_object_name: string | null
    minio_presigned_url: string
}

export interface BookWithChapters {
    book: AudiobookBook
    chapters: AudiobookChapterQueryResult[]
    available_voices: string[]
}
